has-selector.js

```diff
@@ -82,39 +98,50 @@
         case "}":
             if (inAtBlck) inAtBlck--;
             if (inStyles) inStyles--;
-            i = f.index + 1;
+            outerBgn = f.index + 1;
             break;
 
         case ",":
-            if (!inParens && !inStyles && !inAtSlct && !inAtBlck) i = f.index + 1;
+            if (!inStyles && !inAtSlct && !inAtBlck) outerBgn = f.index + 1;
+            break;
+
+        case "+":
+        case "~":
+        case " ":
+        case ">":
+            if (inPseudo) inPseudo--;
+            break;
+
+        case ":":
+            if (!inPrnths && !inStyles && !inAtSlct && !inAtBlck && !inPseudo) outerEnd = f.index;
+            if (!inPrnths && !inStyles && !inAtSlct && !inAtBlck && !inPseudo) inPseudo++;
             break;
 
-        case ":has":
-            h = f.index;
-            inHasDec++;
+        case "has":
+            if (inPseudo) inHasDec = inPrnths + 1;
+            if (inPseudo) innerBgn = f.index - 1;
         }
 
         // This function creates a unique alias, replaces the :has declaration with it,
         // and records the :has declaration's outer and inner element selectors.
         // Without it, it's content would be nested within  switch statement.
-        function addAlias() {
+        function addHasAlias() {
             var A = "abcdefghijklmnopqrstuvwxyz"; // valid characters for the beginning of a class name
             var n = Math.random() * A.length | 0; // index determining which of the above characters to use
-            var c = A[n] + basis.toString(36);    // unique class name
-            var L = 5;                            // the number of characters in the string ":has"
-
-            basis += 1;
-            alias.push(c);
-            r.lastIndex = h;
-            foundAny = true;
-            outer.push(t.slice(i, h));
-            inner.push(t.slice(h + L, f.index));
-            t = t.slice(0, h) + "." + c + t.slice(f.index + 1);
+            var a = A[n] + unique.toString(36);   // unique class name
+            var L = 5;                            // the length of the string ":has"
+
+            unique += 1;
+            alias.push(a);
+            r.lastIndex = innerBgn;
+            outer.push(styleTxt.slice(outerBgn, outerEnd));
+            inner.push(styleTxt.slice(innerBgn + L, innerEnd - 1));
+            styleTxt = styleTxt.slice(0, innerBgn) + "." + a + styleTxt.slice(innerEnd);
         }
 
-        if (foundAny) {
-            s.appendChild(document.createTextNode(t));
-            link.parentNode.replaceChild(s, link);
+        if (xhr.responseText !== styleTxt) {
+            newStyle.appendChild(document.createTextNode(styleTxt));
+            oldElem.parentNode.replaceChild(newStyle, oldElem);
```

Comment:
A big YES to these two lines, much easier to understand than the previous revision.